lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,217 @@
+  # @param {Editor} editor: editor object.
+  # @return {String}: the current word.
+  extractWord = (editor) ->
+    session = editor.getSession()
+    wordRange = getCurrentWordRange(editor)
+    return session.getTextRange(wordRange)
+
+  # Returns Range object that describes the current word position.
+  # @param {Editor} editor: editor object.
+  # @return {Range}: current word range.
+  getCurrentWordRange = (editor) ->
+    session = editor.getSession()
+    row = editor.getCursorPosition().row
+    col = editor.getCursorPosition().column
+    return session.getWordRange(row, col)
+
+  # Convert an array of string to popup-eligible structure.
+  # @param {Array} corrections: array of strings with substitution options.
+  # @return {Array}: array of JSONs, actually.
+  convertCorrectionList = (corrections) ->
+    return ({caption: item, value: item} for item in corrections)
+
+
+  # Object contains init, show and detach functions for popup
+  # and some routines for inner use.
+  PopupManager =
+    # Bindings for editor keys with PopupManager functions.
+    commands:
+      "Esc": ->
+        PopupManager.detach()
+      "Up": ->
+        PopupManager.goTo("up")
+      "Down": ->
+        PopupManager.goTo("down")
+      "Return": ->
+        PopupManager.insertCorrection()
+
+    # Initial setup for popup.
+    # @param {Editor} editor: editor object.
+    init: (editor) ->
+      @editor = editor
+      @activated = false
+
+      HashHandler = require("ace/keyboard/hash_handler").HashHandler
+      AcePopup = require("ace/autocomplete/popup").AcePopup
+
+      @popup = new AcePopup(document.body)
+      @popup.setTheme(editor.getTheme())
+      @popup.setFontSize(editor.getFontSize())
+      @popup.on("click", (e) =>
+        @insertCorrection()
+        e.stop()
+      )
+
+      @session = editor.getSession()
+
+      @keyboard = new HashHandler()
+      @keyboard.bindKeys(@commands)
+
+      return
+
+    # Show popup in editor.
+    # @param {Array} options: list of corrections for the current word.
+    show: (options) ->
+      PopupManager.detach()
+
+      @options = options
+      @popup.setData(options)
+
+      @activated = true
+
+      position = @editor.getCursorPosition()
+      @base = @session.doc.createAnchor(position.row, position.column)
+
+      @editor.keyBinding.addKeyboardHandler(@keyboard)
+      @editor.on("change", -> PopupManager.detach())
+      @editor.on("changeSelection", -> PopupManager.onChangeSelection())
+      @editor.on("blur", -> PopupManager.detach())
+      @editor.on("mousedown", -> PopupManager.detach())
+      @editor.on("mousewheel", -> PopupManager.detach())
+      @editor.$blockScrolling = Infinity
+
+      renderer = @editor.renderer
+      lineHeight = renderer.layerConfig.lineHeight
+      pos = renderer.$cursorLayer.getPixelPosition(this.base, true)
+      pos.left -= @popup.getTextLeftOffset()
+      rect = @editor.container.getBoundingClientRect()
```

Comment:
are you trying to calculate popup position in the browser client coordinates? Is it possible to use coordinates relative to the editor container (roughly, what is returned by `getPixelPosition`, I believe) ?